Test/Integration/Category/CategoryCacheTest.php

```diff
@@ -0,0 +1,341 @@
+            'catalogsearch_fulltext'
+        ];
+
+        foreach ($indexerCodes as $indexerCode) {
+            $indexerRegistry->get($indexerCode)->reindexAll();
+        }
+    }
+
+    /**
+     * Selectively refresh the FPC cache (must be done at intervals)
+     * Warm the cache via MISS tests but only reset the cache once per MISS test
+     * Due to data provider test methods can be called multiple times
+     */
+    protected function resetCache(string $testMethod): void
+    {
+        if (!in_array($testMethod, self::$cacheResets)) {
+            $this->cacheManager->clean(['full_page']);
+            self::$cacheResets[] = $testMethod;
+        }
+    }
+
+    /**
+     * @dataProvider getCategoryProvider
+     * @depends      testFullPageCacheAvailable
+     * @magentoConfigFixture current_store system/full_page_cache/caching_application 1
+     * @magentoConfigFixture current_store algoliasearch_advanced/advanced/prevent_backend_rendering 0
+     * @magentoConfigFixture current_store algoliasearch_instant/instant/replace_categories 1
+     * @magentoCache full_page enabled
+     */
+    public function testCategoryPlpMissBackendRenderOn(int $categoryId, string $name, bool $hasProducts): void
+    {
+        $replace = $this->config->getValue('algoliasearch_instant/instant/replace_categories', ScopeInterface::SCOPE_STORE);
+        $this->assertEquals(1, $replace,"Replace categories must be enabled for this test.");
+
+        $this->resetCache(__METHOD__);
+        $this->dispatchCategoryPlpRequest($categoryId);
+        $response = $this->getResponse();
+        $this->assertEquals(200, $response->getHttpResponseCode(), 'Request failed');
+        $this->assertEquals(
+            'MISS',
+            $response->getHeader('X-Magento-Cache-Debug')->getFieldValue(),
+            "expected MISS on category {$name} id {$categoryId}"
+        );
+        $this->assertContains(
+            'FPC',
+            explode(',', $response->getHeader('X-Magento-Tags')->getFieldValue()),
+            "expected FPC tag on category {$name} id {$categoryId}"
+        );
+
+        if ($hasProducts) {
+            $this->assertMatchesRegularExpression('/<div.*class=.*products-grid.*>/', $response->getContent(), 'Backend content was not rendered.');
+        }
+    }
+
+    /**
+     * @dataProvider getCategoryProvider
+     * @depends      testCategoryPlpMissBackendRenderOn
+     * @magentoConfigFixture current_store system/full_page_cache/caching_application 1
+     * @magentoConfigFixture current_store algoliasearch_advanced/advanced/prevent_backend_rendering 0
+     * @magentoConfigFixture current_store algoliasearch_instant/instant/replace_categories 1
+     * @magentoCache full_page enabled
+     */
+    public function testCategoryPlpHitBackendRenderOn(int $categoryId, string $name): void
+    {
+        $replace = $this->config->getValue('algoliasearch_instant/instant/replace_categories', ScopeInterface::SCOPE_STORE);
```

Comment:
Maybe the two first lines of each test can be factorized ?

Reply:
Yup definitely some redundancy here we can streamline. I was mainly making sure an accidental copy/paste on an annotation doesn't cause the test to fail for the wrong reason.